diff --git a/org/mvel2/compiler/AbstractParser.java b/org/mvel2/compiler/AbstractParser.java
index 1f10748e9094d..f494e7e8bb2bd 100644
--- a/org/mvel2/compiler/AbstractParser.java
+++ b/org/mvel2/compiler/AbstractParser.java
@@ -151,7 +151,6 @@ public class AbstractParser implements Parser, Serializable {
       CLASS_LITERALS.put("ClassLoader", ClassLoader.class);
       CLASS_LITERALS.put("Runtime", Runtime.class);
       CLASS_LITERALS.put("Thread", Thread.class);
-      CLASS_LITERALS.put("Compiler", Compiler.class);
       CLASS_LITERALS.put("StringBuffer", StringBuffer.class);
       CLASS_LITERALS.put("ThreadLocal", ThreadLocal.class);
       CLASS_LITERALS.put("SecurityManager", SecurityManager.class);
@@ -2825,3 +2824,4 @@ public class AbstractParser implements Parser, Serializable {
     this.debugSymbols = (this.pCtx = pCtx).isDebugSymbols();
   }
 }
+/* @generated */
\ No newline at end of file
